### DISCUSSION TITLE
SpanButton seems inverted?

### DISCUSSION THREAD

**Discussion by andreasp79** [HomeSpan/HomeSpan#237] [Q&A]
I made a simple circuit with a normally open button, and created a programmable switch using code from example 16 - ProgrammableSwitches. The button events HomeSpan is sending to HomeKit are inverted - it keeps sending a long press as if the button is kept pushed when it isn't. If I keep the button pushed, then momentarily release it, then it sends a single press event. 

Is this normal? I was expecting this to work the opposite way. See my sketch below, the "14 pressed" serial log prints as expected only when I keep the button pressed. The Homekit pointer is simply implementing the Programmable Switch on pin 14, exactly as example 16 shows. 

```
#include "homekit.h"

Homekit* pHome;

void setup() 
{
  Serial.begin(115200);

  pinMode(14, INPUT);

    pHome = new Homekit();
    pHome->Initialize();
}

void loop() 
{
  if (digitalRead(14))
     Serial.println("14 pressed");
  
  pHome->Update();
}
```

**Comment by HomeSpan** [comment]
If you implemented the full Programmable Switch example, that already includes the creation of SpanButtons  - you don't need to set the pinMode (or do anything else to initialize a button).  Also SpanButtons are based on a pushbuttons that connect a pin to ground when closed.  The code in your loop() is doing the opposite and would only work if you connected the pin to +3V through a button, in which case SpanButton will not function correctly.   How did you configure you physical button - to ground or to +3V?

**Comment by andreasp79** [comment]
i configured it to 3.3v, as described here https://www.arduino.cc/en/Tutorial/BuiltInExamples/Button

Is there an example of how i would wire everything to make it go to 3v when button is untouched and ground when button is pressed? Obviously i'm a beginner, thank you for the help

**Comment by HomeSpan** [comment]
Simply wire one end of the button to the pin you chose and wire the other end of the button to a pin that says GND.  Then run the Programmable Button example exactly as is -  no modifications except for the line where you provide the pin number of your button.  Once you get that working as expected than you can start modifying, but at least you'll have a baseline to start.

**Comment by andreasp79** [comment]
Thanks again, works now. And thanks you for writing Homespun. Very easy to use and it works very well (better than most commercial accessories that never want to connect the first time)

**Comment by HomeSpan** [comment]
Glad it worked out!
